subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestClassProcessor.java

```diff
@@ -83,21 +86,25 @@ public void stopNow() {
         @Override
         public void execute(String testClassName) {
             Class<?> klass = loadClass(testClassName);
-            if (isTopClass(klass)) {
-                testClasses.add(klass);
+            if (isInnerClassAnnotatedWithNested(klass) || isInnerClassInsideEnclosedRunner(klass)) {
```

Comment:
The nested annotation shouldn't need any handling here, as it is about *inner* classes.

The enclosed runner is used with static nested classes, so the name of the method here is wrong. It is not an inner class.

Reply:
I think we need to handle `@Nested` here - according to [JUnit 5 userguide](https://junit.org/junit5/docs/current/user-guide/#writing-tests-nested):

```
class TestingAStackDemo {

    Stack<Object> stack;

    @Test
    void isInstantiatedWithNew() {
        new Stack<>();
    }

    @Nested
    class WhenNew {

        @BeforeEach
        void createNewStack() {
            stack = new Stack<>();
        }

        @Test
        void isEmpty() {
            assertTrue(stack.isEmpty());
        }

    }
}
```

We have to ignore `TestingAStackDemo$WhenNew`, otherwise it might be processed twice by two worker processes.

As for `isInnerClassInsideEnclosedRunner`, I agree with you. I was a bit confused here. Changed to `isNestedClassInsideEnclosedRunner`.

Reply:
WhenNew won't be run twice. It's an inner class which cannot be run on its own. JUnit won't pick it up.

Reply:
I think we're discussing JUnit 5. I'm pretty sure it will. If we pass `TestingAStackDemo$WhenNew` class to JUnit 5, it will be run although it can't be initialized directly (I guess JUnit 5 fetches its enclosing class then initializes it). That's why I added this test [here](https://github.com/gradle/gradle/commit/88b010301ff55c7576c0e58469c133fc719bdaa6#diff-4f366fdd84cf8f0d203cc219e3e2f49f).

Reply:
If that's the case, then we should change the behavior to never pass inner classes.

Reply:
I don't like the special cases here either, but I think it's necessary because JUnit 5 is a general platform, we shouldn't make any assumptions and "detain" any classes privately.

Reply:
I'm okay with this solution, but @marcphilipp I'd like to hear your opinion.